### DISCUSSION TITLE
Take advantage of GitHub Actions to auto-update Workshop

### DISCUSSION THREAD

**Discussion by Cheatoid** [vurvdev/VExtensions#33] [Ideas]
I have started exploring GitHub Actions some day ago... and I'm actually already using this tech (currently in a private repo), it is really nice.
So, there was a suggestion, it can be used to do just about anything, trigger-based or scheduled, guarded with conditions such as "only do this when a commit is pushed to *master* branch", "do this when new issue is created, and the title contains XYZ"...
It goes way beyond just that, it can also: "run Lua/Python/JavaScript/C#/*you-name-it* as part of the action", "perform Lua code analysis/linting", "send a message in Discord"... [creativity is the limitation here<kbd>*</kbd>]

The problem is, I am not 100% sure if we are allowed to redistribute `gmad` and `gmpublish` binaries, these will have to be provided in some way (obviously), an Action will have to either download them from some "secret" Facepunch endpoint (if there is any), or, if their redistribution is allowed then we can just upload them as a part of the repo easy-peasy...
Then we can execute these programs within an Action as part of the workflow, therefore an addon can be auto-updated on Workshop.

Moreover, it is also possible to automatically create a [GitHub release](https://github.com/Vurv78/VExtensions/releases) as well, I would prefer to have a tagged-commit condition for it...

The idea is: Everything can be automated basically. Even for servers which use git, they can be setup to run `git pull` as soon as the new commit is pushed out (this can be securely achieved with "push notifications" or via webhooks, similarly to how a message can be sent to Discord bot from an Action...).

<kbd>*</kbd> Here's [the random cool example](https://github.com/DoubleGremlin181/DoubleGremlin181), it is the live Tic-Tac-Toe game powered by Actions! **Anyone** (with GitHub account) is able to make the move (click the tile and just proceed to create an "issue", don't touch the generated text), when the "issue" is created, it will trigger an Action which will run the game logic and update the readme automatically to reflect the new board state... Brilliant, right?

**Comment by Cheatoid** [comment]
Oh yeah I forgot to mention the important note: any "secrets" such as Steam API key or anything password-sensitive **must be** configured with [GitHub Secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets) page, in order for it to be secure because sharing such a secret would be sad, if it were publicly committed.
Although, GitHub support will be able to help in such a case, and they are able to remove the commit if necessary, to help prevent info leak.

----

Also, an Action can be configured to run across multiple operating systems including Windows, Linux and macOS 😄

**Comment by Cheatoid** [comment]
Another cool idea I have in mind and I would like to share it here (so I don't forget it):
- [ ] Make an Action which would run through Lua file(s) that contain all E2 descriptions, automatically generate Markdown (documentation) files and pushing it onto the "`docs`" (new) branch (or perhaps a new directory in the root of the repo).
    * This would enable us to have Markdown docs automatically generated as soon as new E2 functions are added.
    * Perhaps also parse files which contain `e2function` (DSL) in order to extract parameter names... Juicy. 😄

----

If you have any ideas on what else to do with Actions, throw your ideas in here, I wanna play with Actions! 🚀

**Reply by vurvdev** [reply]
It'd be cool if we could get a lua linter for gmod, and I think metaconstruct already did that. We'd need to adjust it for e2functions and stuff tho

**Reply by vurvdev** [reply]
https://github.com/FPtje/GLuaFixer or https://github.com/Metastruct/gtravis
